### ISSUE TITLE
Update whereami utility to be cloud-agnostic

### ISSUE THREAD

**Issue by theemadnes** [GoogleCloudPlatform/kubernetes-engine-samples#207]
@knee-berts brought up an excellent point - with Anthos being multi-cloud, `whereami` should be able to return the same payload regardless of which supported platform it's running on. Currently investigating implementation approaches with him.

**Comment by theemadnes** [comment]
potential platforms include GCP, AWS, Azure, etc

**Comment by theemadnes** [comment]
currently `whereami` references the GCE metadata endpoint to pull some info, so effectively we need to come up with a non-GCP alternative

**Comment by theemadnes** [comment]
Also potentially in scope: non-cloud platforms like bare metal & vSphere.

**Comment by askmeegs** [comment]
Hi Alex, I agree this is a good idea that could help with Anthos multicloud demos. 
What happens today when this app runs on AWS or Azure? Does it throw an error trying to call the GCE metadata endpoint? 

Because there are two options -- 
1) Fail gracefully if the app can't reach the GCE metadata server. That's what we're doing with the bigger sample apps today. The region just shows up as "unknown" in the frontend. https://github.com/GoogleCloudPlatform/bank-of-anthos/blob/master/src/frontend/frontend.py#L543 

2) Add AWS / Azure specific code to detect what cloud provider the user is running in, and call the appropriate server to fetch the region/zone.  But this is a bit of a rabbit hole - ie. will you also cover baremetal, kind, and all the other Anthos supported environments?  Arguably then, this app would not belong in this repo (GKE on GCP samples) but rather  https://github.com/GoogleCloudPlatform/anthos-samples which covers the non GCP environments. 

It may be also worth asking the Anthos multicloud team if they're using any in-house apps to demo the product / see if something similar has ever been done.

**Comment by theemadnes** [comment]
incremental updated, but the PR i just put in will move the node name source from GCE metadata to Downward API, which makes it work on any K8s environment

**Comment by askmeegs** [comment]
Thanks Alex! We might try this workaround in the Bank of Anthos and Online Boutique sample apps.

**Comment by theemadnes** [comment]
cool. happy to contribute those changes if you're looking to offload some work. LMK.

**Comment by theemadnes** [comment]
also @askmeegs the latest PR doesn't totally solve the incomplete payload issue - it just makes the node_name field work, which hopefully is sufficient for most use cases. the fields that won't work outside of GCP after this PR are as follows:

- `project_id` of course
- `zone` zones of course are also applicable to other cloud providers, but would require cloud-specific implementations
- `cluster_name` this one is tricky in large part because within a given K8s cluster, the cluster doesn't "know" its own name;  I've though about potentially using Fleets to address this, but it (at first blush, at least) seems difficult to implement because within that cluster, the `gke-connect` namespace has a few references to the cluster name (in the `deployment` and corresponding `pod` specs, but i'd need to slice up some string and I need to do some more testing to make it work

**Comment by theemadnes** [comment]
TLDR is, if I'd need to write some `makefile` or something similar to make this work, it's probably not worth the added complexity for users seeing as `whereami` was written to be stupid-simple to spin up, and the kustomize stuff might break. i'll continue investigating though to see if we can make it easy.
